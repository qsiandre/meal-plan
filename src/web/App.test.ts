import { expect, describe, it, beforeAll } from "bun:test";
import puppeteer, { Browser, Page } from "puppeteer-core";

async function navigateTo(browser: Browser | null, url: string): Promise<Page> {
  const page = await browser?.newPage();
  if (page == null) {
    throw Error("Couldn't navigate to page");
  }
  await page.goto(url);
  return page;
}

async function parse(page: Page, ...urls: string[]): Promise<void> {
  for (const url of urls) {
    await page.type("input[placeholder='Enter URL']", url);
    await page.click("button >>> ::-p-text('+')");
  }
  await page.click("button >>> ::-p-text(Next)");
}

describe("App", () => {
  let browser: Browser | null = null;
  beforeAll(async () => {
    browser = await puppeteer.launch({
      headless: "new",
      executablePath: Bun.env["CHROME_PATH"],
      args: ["--no-sandbox"],
    });
  });

  it("extracts multiple recipes from TJ's website", async () => {
    const page = await navigateTo(browser, "http:/localhost:3000");
    await parse(
      page,
      "https://www.traderjoes.com/home/recipes/veggie-chickpea-hash",
      "https://www.traderjoes.com/home/recipes/smoked-salmon-pasta-salad"
    );
    const veggie = await page.waitForSelector(
      "div >>> ::-p-text(Veggie Chickpea Hash)",
      { timeout: 10000 }
    );
    const veggieText = await veggie?.evaluate((e) => (e as any).innerText);
    expect(veggieText).toMatchSnapshot();
    const pasta = await page.waitForSelector(
      "div >>> ::-p-text(Smoked Salmon Pasta Salad)",
      { timeout: 10000 }
    );
    const pastaText = await pasta?.evaluate((e) => (e as any).innerText);
    expect(pastaText).toMatchSnapshot();
  });

  it.only("edits ingredients from a recipe", async () => {
    const page = await navigateTo(browser, "http:/localhost:3000");
    await parse(
      page,
      "https://www.traderjoes.com/home/recipes/veggie-chickpea-hash"
    );
    await page.waitForSelector("button >>> ::-p-text(Save)", {
      timeout: 10000,
    });
    await page.type("textarea >>> ::-p-text(Eggs)", "Edit ");
    await page.click("button >>> ::-p-text(Save)");
    await page.click("button >>> ::-p-text(Next)");
    const itemText = await page.$eval(
      "li >>> ::-p-text(Eggs)",
      (e) => (e as any).innerText
    );
    expect(itemText).toBe("Edit 4 TJâ€™s Eggs");
    const list = await page.$eval("ul", (e) => (e as any).innerText);
    expect(list).toMatchSnapshot();
  });
});
