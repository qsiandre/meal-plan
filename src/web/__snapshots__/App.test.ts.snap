// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`App extracts multiple recipes from TJ's website 1`] = `
"{
    "url": "https://www.traderjoes.com/home/recipes/veggie-chickpea-hash",
    "title": "Veggie Chickpea Hash",
    "image": "https://www.traderjoes.com/content/dam/trjo/context-images/60205-greek-chickpeas-2-pdp.jpg/jcr:content/renditions/cq5dam.web.1280.1280.jpeg",
    "serves": "Serves 1-4",
    "time": "Time 25 mins"
}"
`;

exports[`App extracts multiple recipes from TJ's website 2`] = `
"{
    "url": "https://www.traderjoes.com/home/recipes/smoked-salmon-pasta-salad",
    "title": "Smoked Salmon Pasta Salad",
    "image": "https://www.traderjoes.com/content/dam/trjo/context-images/74547-org-striped-farfalline-pdp.jpg/jcr:content/renditions/cq5dam.web.1280.1280.jpeg",
    "serves": "Serves 4-6",
    "time": "Time 40 mins"
}"
`;

exports[`App edits ingredients from a recipe 1`] = `
"1 can TJ’s Greek Chickpeas with Parsley & Cumin
1 TJ’s Red Onion, chopped
1 TJ’s Red Bell Pepper, chopped
1 teaspoon TJ’s Smoked Paprika
A few shakes of TJ’s 21 Seasoning Salute
4 TJ’s Eggs Edit"
`;
